docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+consensus. Because the event system is very cross cutting, the behavior and
+performance of the event publication and subscription system has huge impacts
+for all of Tendermint.
+
+The subscription service is exposed over the RPC interface, but also powers
+the indexing (e.g. to an external database,) and is the mechanism by which
+`BroadcastTxCommit` is able to wait for transactions to land in a block.
+
+The current pubsub mechanism relies on a couple of buffered channels,
+primarily between all event creators and subscribers, but also for each
+subscription. The result of this design is that, in some situations with the
+right collection of slow subscription consumers the event system can put
+backpressure on the consensus state machine and message gossiping in the
+network, thereby causing nodes to lag.
+
+Improvements
+~~~~~~~~~~~~
+
+The current system relies on implicit queues built by the buffered channels,
+and though threadsafe, can force all activity within Tendermint to serialize,
+which does not need to happen. Additionally, timeouts for subscription
+consumers related to the implementation of the RPC layer, may complicate the
+use of the system.
+
+References
+~~~~~~~~~~
+
+- Legacy Implementation
+  - `publication of events <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L333-L345>`_ 
+  - `send operation <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L489-L527>`_ 
+  - `send loop <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L381-L402>`_
+- Related RFCs 
+  - `RFC 002: IPC Ecosystem <./rfc-002-ipc-ecosystem.md>`_ 
+  - `RFC 003: Performance Questions <./rfc-003-performance-questions.md>`_ 
+
+Discussion
+----------
+
+Changes to Published Events
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As part of this process, the Tendermint team should do a study of the existing
+event types and ensure that there are viable production use cases for
+subscriptions to all event types. Instinctively it seems plausible that some
+of the events may not be useable outside of tendermint, (e.g. ``TimeoutWait``
+or ``NewRoundStep``) and it might make sense to remove them. Certainly, it
+would be good to make sure that we don't maintain infrastructure for unused or
+un-useful message indefinitely.
+
+Blocking Subscription
+~~~~~~~~~~~~~~~~~~~~~
+
+The blocking subscription mechanism makes it possible to have *send*
+operations into the subscription channel be un-buffered (the event processing
+channel is still buffered.) In the blocking case, events from one subscription
+can block processing that event for other non-blocking subscriptions. The main
+case, it seems for blocking subscriptions is ensuring that a transaction has
+been committed to a block for ``BroadcastTxCommit``. Removing blocking
+subscriptions entirely, and potentially finding another way to implement
+``BroadcastTxCommit``, could lead to important simplifications and
```

Comment:
My general instinct here is that `BroadcastTxCommit` is unlikely to be all that useful as is. Since Txs can be low priority and therefore never make it into a block, or wait for a very long time, clients are very likely to timeout waiting for a response in that scenario. 

A simplification would likely put the onus on clients to poll the state to see if their Tx was included in a block or booted from the mempool.

Reply:
Looks like the cosmos SDK has noted this problem:
https://github.com/cosmos/cosmos-sdk/blob/260c5728ab3b759788be68f04dd5d0ec07be0b78/client/broadcast.go#L88

Reply:
Yes, agreed that this change would require a lot of coordination. The part I think is worth memorializing in the RFC is the desire to move in that direction (about which we seem more-or-less in accord).

Reply:
I also concur with this. The events between getting a transaction in the mempool and having it committed in a block are distinct enough that although users may tend to want a "fire and wait a reasonable amount of time for it to be confirmed", I think it makes sense for Tendermint to have them as separate actions (submit a transaction, poll for confirmation) and rely on the front-end client to do the lifting. If we are going down this route, I think there are few adjustments we should consider making to improve the visibility of the mempool (some of these points stem from the conversation we had at the office hours):
1. RPC clients should be made aware of the TTL of a transaction to know the upper bound with which they can expect a transaction to be delivered by before having to retry. This is however a little bit more complicated than what it may seem because the TTL settings are node specific - another node could hold that transaction for eternity and propose it in a block after that time. At the end of the day it's up to the application (in most cases the SDK) to handle retry and concurrent transaction logic. We can just do the best we can to support that.
2. RPC clients could be made aware of what the minimum priority that was required to put a transaction in the last block was. AFAIK, there's no way for clients to know if the priority of their submitted transaction is anywhere near what the priority of other transactions are. 
3. RPC clients could be made aware of what transactions were dropped. This would be helpful so that clients know they need to resend their transactions but once again this is a node-specific element and some nodes could support enormous mempools in which nothing gets dropped. There are some circumstances where the dropping of a transaction is system-wide which happens when transactions get re-check-tx'ed.

Reply:
>RPC clients should be made aware of the TTL of a transaction to know the upper bound with which they can expect a transaction to be delivered by before having to retry. This is however a little bit more complicated than what it may seem because the TTL settings are node specific - another node could hold that transaction for eternity and propose it in a block after that time

This definitely adds a bit of complexity since the guarantee we can make is a bit weaker. I do think it may be worth considering changing our mempool TTL to be a consensus param and enforcing it network-wide. Tx could then be submitted with a `HeightSubmittedAt` which would be gossiped along with the Tx so nodes would all know when to evict it. 

Worth considering these options if we do:

https://github.com/tendermint/tendermint/issues/6958